packages/ghost/public/ghost-worker.js

```diff
@@ -0,0 +1,127 @@
+const DESTROY = 5000;
+const TIMEOUT = 100;
+
+const virtualFS = new Map();
+
+let ready = false;
+let scope = "";
+
+const getMimeType = (path) => {
+    const mimeMap = {
+        ".css": "text/css",
+        ".js": "application/javascript",
+        ".json": "application/json",
+        ".jsonp": "application/javascript",
+        ".png": "image/png",
+        ".jpg": "image/jpeg",
+        ".html": "text/html",
+        ".svg": "image/svg+xml",
+        ".woff": "font/woff",
+        ".woff2": "font/woff2",
+        ".ttf": "font/ttf",
+    };
+    const ext = path.slice(path.lastIndexOf(".")).split("?")[0];
+    return mimeMap[ext] || "text/plain";
+};
+
+self.addEventListener("install", (event) => {
+    console.log("[GHOST] Installing...");
+    event.waitUntil(self.skipWaiting());
+});
+
+self.addEventListener("activate", (event) => {
+    console.log("[GHOST] Activating...");
+    event.waitUntil(clients.claim());
+});
+
+self.addEventListener("message", (event) => {
+    if (event.data.type === "CREATE") {
+        scope = event.data.scope;
+        const files = Object.entries(event.data.files);
+        for (const [path, content] of files) {
+            virtualFS.set(path, content);
+        }
+        console.log(`[GHOST] Serving ${files.length} files from ${scope}`);
+        ready = true;
+    }
+    if (event.data.type === "DESTROY") {
+        virtualFS.clear();
+        console.log("[GHOST] Destroyed filesystem");
+        ready = false;
+    }
+});
+
+self.addEventListener("fetch", (event) => {
+    console.log("[GHOST] Intercepting...");
+    const url = new URL(event.request.url);
+    const isSameOrigin = url.origin === self.location.origin;
+    const scoped = scope.endsWith("/") ? scope : scope + "/";
+
+    // Only handle same-origin requests
+    if (isSameOrigin) {
+        if (event.request.method === "GET" && url.pathname.startsWith(scoped)) {
```

Comment:
get requests include the session key, this might be a concern. How does the interception work, can a script that's loaded through the worker makes requests using their own shipped networking library ? Have you tried exploiting this with malicious archives ?

Reply:
I am experimenting with an option that injects the actual dataset and sandboxes the entire visualization. This would limit the file size, which is a drawback, but likely make it safe fyi.